packages/lib/src/vis/tiles/api.ts

```diff
@@ -0,0 +1,50 @@
+import type { NdArray } from 'ndarray';
+import type { Vector2 } from 'three';
+
+import type { TextureSafeTypedArray } from '../heatmap/models';
+import type { Size } from '../models';
+
+export function getLevelsSize(
+  size: Size,
+  tileSize: Size,
+  evenLevelSize = false
+): Size[] {
+  const levelsSize: Size[] = [];
+  do {
+    const scale = 2 ** levelsSize.length;
+    if (evenLevelSize) {
+      levelsSize.push({
+        width: Math.floor(size.width / scale),
+        height: Math.floor(size.height / scale),
+      });
+    } else {
+      levelsSize.push({
+        width: 1 + Math.floor((size.width - 1) / scale),
+        height: 1 + Math.floor((size.height - 1) / scale),
+      });
+    }
+  } while (
+    levelsSize[levelsSize.length - 1].width > tileSize.width &&
+    levelsSize[levelsSize.length - 1].height > tileSize.height
+  );
+  return levelsSize;
+}
+
+export abstract class TilesApi {
+  public readonly tileSize: Size;
+  protected levelsSizeArray: Size[];
+
+  public constructor(tileSize: Size, levelsSize: Size[]) {
+    this.tileSize = tileSize;
+    this.levelsSizeArray = [...levelsSize];
+  }
+
+  public get levelsSize() {
+    return [...this.levelsSizeArray];
```

Comment:
I wouldn't bother with this getter and with copying the array in the constructor. Just mark it as `readonly` and TS should prevent anyone from modifying it (I forgot about this feature when you asked me about freezing objects a while back).